File: bridge/MCPStdioBridge.swift
//
// MCPStdioBridge.swift — zero-dependency stdio-to-HTTP bridge for MCP servers.
//
// Build:   swiftc -O -o mcp-stdio-bridge MCPStdioBridge.swift
//
// Design baseline: mootx01-ce apps/mootx01/rust/src/commands/proxy.rs

import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

// MARK: - Configuration

let defaultURL = "http://127.0.0.1:8765/mcp"
let defaultTimeout: TimeInterval = 3600
let defaultMaxConcurrent = 16
let maxLineBytes = 4 * 1024 * 1024

struct Config {
    var url: String
    var timeout: TimeInterval
    var maxConcurrent: Int
    var logFile: String?
}

func parseArgs() -> Config {
    var config = Config(url: defaultURL, timeout: defaultTimeout,
                        maxConcurrent: defaultMaxConcurrent, logFile: nil)
    let args = Array(CommandLine.arguments.dropFirst())
    var i = 0
    while i < args.count {
        switch args[i] {
        case "--url" where i + 1 < args.count:
            config.url = args[i + 1]; i += 2
        case "--timeout" where i + 1 < args.count:
            config.timeout = TimeInterval(args[i + 1]) ?? defaultTimeout; i += 2
        case "--max-concurrent" where i + 1 < args.count:
            config.maxConcurrent = Int(args[i + 1]) ?? defaultMaxConcurrent; i += 2
        case "--log-file" where i + 1 < args.count:
            config.logFile = args[i + 1]; i += 2
        case "-h", "--help":
            log("Usage: mcp-stdio-bridge [--url URL] [--timeout SECS] [--max-concurrent N] [--log-file PATH]")
            exit(0)
        default:
            log("unknown option: \(args[i])"); exit(1)
        }
    }
    return config
}

// MARK: - Logging (stderr + optional file, never stdout)

var logFileHandle: FileHandle?
let logLock = NSLock()

func setupLogFile(_ path: String) {
    FileManager.default.createFile(atPath: path, contents: nil)
    logFileHandle = FileHandle(forWritingAtPath: path)
    logFileHandle?.seekToEndOfFile()
}

func log(_ msg: String) {
    let ts = ISO8601DateFormatter().string(from: Date())
    let line = "[\(ts)] mcp-stdio-bridge[\(ProcessInfo.processInfo.processIdentifier)]: \(msg)\n"
    let data = Data(line.utf8)
    FileHandle.standardError.write(data)
    logLock.lock()
    logFileHandle?.write(data)
    logLock.unlock()
}

// MARK: - Session state (thread-safe)

let sessionLock = NSLock()
var sessionID: String?

func getSessionID() -> String? {
    sessionLock.lock(); defer { sessionLock.unlock() }
    return sessionID
}

func setSessionID(_ sid: String) {
    sessionLock.lock(); defer { sessionLock.unlock() }
    let old = sessionID
    sessionID = sid
    if old != sid {
        log("session: \(old ?? "nil") -> \(sid)")
    }
}

// MARK: - Stdout serialization

let stdoutLock = NSLock()

func writeFrame(_ data: Data) {
    stdoutLock.lock(); defer { stdoutLock.unlock() }
    var out = data
    out.append(0x0A)
    FileHandle.standardOutput.write(out)
}

// MARK: - SSE parsing

func extractSSEData(_ body: Data) -> [Data] {
    guard let text = String(data: body, encoding: .utf8) else { return [body] }
    var frames: [Data] = []
    for line in text.components(separatedBy: "\n") {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        if trimmed.hasPrefix("data: ") {
            let payload = String(trimmed.dropFirst(6))
            if !payload.isEmpty, let d = payload.data(using: .utf8) {
                frames.append(d)
            }
        }
    }
    return frames.isEmpty ? [body] : frames
}

// MARK: - JSON-RPC id and method extraction

func requestID(_ line: String) -> String? {
    guard let data = line.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let rawID = obj["id"] else { return nil }
    if rawID is NSNull { return nil }
    if let s = rawID as? String {
        if let encoded = try? JSONSerialization.data(withJSONObject: s) {
            return String(data: encoded, encoding: .utf8)
        }
        return nil
    }
    if let n = rawID as? NSNumber { return n.stringValue }
    return nil
}

func requestMethod(_ line: String) -> String? {
    guard let data = line.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let method = obj["method"] as? String else { return nil }
    return method
}

// MARK: - Frame forwarding

func forwardFrame(url: URL, timeout: TimeInterval, line: String,
                  session: URLSession) {
    guard let body = line.data(using: .utf8) else { return }

    let method = requestMethod(line) ?? "unknown"
    let rid = requestID(line) ?? "notification"
    log("-> \(method) id=\(rid)")

    var req = URLRequest(url: url)
    req.httpMethod = "POST"
    req.httpBody = body
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    req.setValue("application/json, text/event-stream", forHTTPHeaderField: "Accept")
    req.timeoutInterval = timeout

    if let sid = getSessionID() {
        req.setValue(sid, forHTTPHeaderField: "Mcp-Session-Id")
    }

    let sem = DispatchSemaphore(value: 0)
    let start = Date()

    let task = session.dataTask(with: req) { data, response, error in
        defer { sem.signal() }
        let elapsed = Date().timeIntervalSince(start)

        if let error = error {
            log("<- \(method) id=\(rid) ERROR [\(String(format: "%.3f", elapsed))s]: \(error.localizedDescription)")
            if let rid = requestID(line) {
                let msg = error.localizedDescription
                    .replacingOccurrences(of: "\\", with: "\\\\")
                    .replacingOccurrences(of: "\"", with: "'")
                let errFrame = Data(
                    "{\"jsonrpc\":\"2.0\",\"id\":\(rid),\"error\":{\"code\":-32603,\"message\":\"bridge: \(msg)\"}}".utf8
                )
                writeFrame(errFrame)
            }
            return
        }

        let httpResp = response as? HTTPURLResponse
        let status = httpResp?.statusCode ?? 0

        if let newSID = httpResp?.value(forHTTPHeaderField: "Mcp-Session-Id") {
            setSessionID(newSID)
        }

        guard let data = data, !data.isEmpty, status != 202 else {
            log("<- \(method) id=\(rid) \(status) empty [\(String(format: "%.3f", elapsed))s]")
            return
        }

        let contentType = httpResp?.value(forHTTPHeaderField: "Content-Type") ?? ""
        log("<- \(method) id=\(rid) \(status) \(data.count)B \(contentType.prefix(30)) [\(String(format: "%.3f", elapsed))s]")

        if contentType.contains("text/event-stream") {
            for frame in extractSSEData(data) {
                if !frame.isEmpty { writeFrame(frame) }
            }
        } else {
            writeFrame(data)
        }
    }
    task.resume()
    sem.wait()
}

// MARK: - Startup health check

func waitForServer(url: URL, maxWait: TimeInterval = 30) -> Bool {
    let ping = Data(#"{"jsonrpc":"2.0","id":"ping","method":"initialize","params":{"protocolVersion":"2025-11-25","capabilities":{},"clientInfo":{"name":"healthcheck","version":"1"}}}"#.utf8)
    let deadline = Date().addingTimeInterval(maxWait)
    var attempt = 0
    while Date() < deadline {
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.httpBody = ping
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json, text/event-stream", forHTTPHeaderField: "Accept")
        req.timeoutInterval = 2
        let sem = DispatchSemaphore(value: 0)
        var ok = false
        let task = URLSession.shared.dataTask(with: req) { _, resp, _ in
            if let http = resp as? HTTPURLResponse, http.statusCode > 0 {
                ok = true
                if let sid = http.value(forHTTPHeaderField: "Mcp-Session-Id") {
                    setSessionID(sid)
                }
            }
            sem.signal()
        }
        task.resume()
        sem.wait()
        if ok { return true }
        attempt += 1
        if attempt == 5 { log("server not up yet at \(url.absoluteString), waiting...") }
        Thread.sleep(forTimeInterval: 0.5)
    }
    return false
}

// MARK: - Main

let config = parseArgs()

if let logPath = config.logFile {
    setupLogFile(logPath)
}

log("starting pid=\(ProcessInfo.processInfo.processIdentifier)")

guard let endpointURL = URL(string: config.url) else {
    log("invalid URL: \(config.url)"); exit(1)
}

guard waitForServer(url: endpointURL) else {
    log("no server answering at \(config.url) after 30s"); exit(1)
}

log("bridging stdio -> \(config.url)")

let sessionConfig = URLSessionConfiguration.default
sessionConfig.httpMaximumConnectionsPerHost = config.maxConcurrent
let urlSession = URLSession(configuration: sessionConfig)

let workerQueue = DispatchQueue(label: "bridge.workers", attributes: .concurrent)
let concurrencyCap = DispatchSemaphore(value: config.maxConcurrent)
let inflightGroup = DispatchGroup()

var buffer = Data()
while true {
    let chunk = FileHandle.standardInput.availableData
    if chunk.isEmpty {
        log("stdin EOF")
        break
    }
    buffer.append(chunk)

    while let nlIndex = buffer.firstIndex(of: 0x0A) {
        let lineData = buffer[buffer.startIndex..<nlIndex]
        buffer.removeSubrange(buffer.startIndex...nlIndex)
        if lineData.isEmpty { continue }
        if lineData.count > maxLineBytes {
            log("frame exceeds \(maxLineBytes) bytes, dropped"); continue
        }
        guard let line = String(data: lineData, encoding: .utf8) else { continue }

        concurrencyCap.wait()
        inflightGroup.enter()
        workerQueue.async {
            forwardFrame(url: endpointURL, timeout: config.timeout,
                         line: line, session: urlSession)
            concurrencyCap.signal()
            inflightGroup.leave()
        }
    }
}

if !buffer.isEmpty, buffer.count <= maxLineBytes,
   let line = String(data: buffer, encoding: .utf8), !line.isEmpty {
    forwardFrame(url: endpointURL, timeout: config.timeout,
                 line: line, session: urlSession)
}

inflightGroup.wait()
log("exiting")

File: bridge/mcp-stdio-bridge.py
#!/usr/bin/env python3
"""
mcp-stdio-bridge — a zero-dependency stdio-to-HTTP bridge for MCP servers.

Replaces npx mcp-remote for local HTTP MCP servers. No Node.js required.
Reads newline-delimited JSON-RPC frames from stdin, POSTs each to the
target HTTP endpoint, writes the response to stdout. Logging goes to
stderr only; stdout carries JSON-RPC frames exclusively.

Concurrency: each frame forwards on its own thread. A slow tool call
(Snapshot, long Shell command) does not block frames behind it. Responses
may interleave out of order; this is legal JSON-RPC (clients correlate
by id).

Failure policy: a failed REQUEST gets a synthesized JSON-RPC error that
echoes the request's id. A failed NOTIFICATION gets nothing (servers must
not reply to notifications). The bridge only exits on stdin EOF or a
stdin read error — never because one call failed.

Handles MCP Streamable HTTP session management (MCP-Session-Id) and SSE
response parsing (text/event-stream with data: lines).

Design baseline: mootx01-ce apps/mootx01/rust/src/commands/proxy.rs

Usage:
    mcp-stdio-bridge.py [--url URL] [--timeout SECONDS] [--max-concurrent N]

Claude Desktop config:
    {
      "mcpServers": {
        "macos-mcp": {
          "command": "python3",
          "args": ["/path/to/bridge/mcp-stdio-bridge.py",
                   "--url", "http://127.0.0.1:8765/mcp"]
        }
      }
    }
"""

import json
import sys
import threading
import time
import urllib.request
import urllib.error
from concurrent.futures import ThreadPoolExecutor

# -- Configuration -----------------------------------------------------------

DEFAULT_URL = "http://127.0.0.1:8765/mcp"
DEFAULT_TIMEOUT = 3600
DEFAULT_MAX_CONCURRENT = 16
MAX_LINE_BYTES = 4 * 1024 * 1024

def parse_args():
    url = DEFAULT_URL
    timeout = DEFAULT_TIMEOUT
    max_concurrent = DEFAULT_MAX_CONCURRENT
    args = sys.argv[1:]
    i = 0
    while i < len(args):
        if args[i] == "--url" and i + 1 < len(args):
            url = args[i + 1]; i += 2
        elif args[i] == "--timeout" and i + 1 < len(args):
            timeout = int(args[i + 1]); i += 2
        elif args[i] == "--max-concurrent" and i + 1 < len(args):
            max_concurrent = int(args[i + 1]); i += 2
        elif args[i] in ("-h", "--help"):
            sys.stderr.write(__doc__); sys.exit(0)
        else:
            sys.stderr.write(f"mcp-stdio-bridge: unknown option: {args[i]}\n"); sys.exit(1)
    return url, timeout, max_concurrent

# -- Session state (thread-safe) ---------------------------------------------

_session_id = None
_session_lock = threading.Lock()

def get_session_id():
    with _session_lock:
        return _session_id

def set_session_id(sid):
    global _session_id
    with _session_lock:
        _session_id = sid

# -- Stdout serialization ----------------------------------------------------

_stdout_lock = threading.Lock()

def write_frame(data: bytes) -> None:
    with _stdout_lock:
        try:
            sys.stdout.buffer.write(data)
            sys.stdout.buffer.write(b"\n")
            sys.stdout.buffer.flush()
        except BrokenPipeError:
            pass

# -- SSE parsing -------------------------------------------------------------

def extract_sse_data(body: bytes) -> bytes:
    """Extract JSON-RPC frames from an SSE text/event-stream response.
    Each 'data: {...}' line is a frame. Multiple frames are joined with newlines."""
    frames = []
    for raw_line in body.split(b"\n"):
        line = raw_line.strip()
        if line.startswith(b"data: "):
            payload = line[6:]  # strip "data: " prefix
            if payload:
                frames.append(payload)
    return b"\n".join(frames) if frames else body

# -- Frame forwarding --------------------------------------------------------

def request_id(line: str):
    try:
        obj = json.loads(line)
    except (json.JSONDecodeError, TypeError):
        return None
    raw_id = obj.get("id")
    if raw_id is None:
        return None
    if isinstance(raw_id, (str, int, float)):
        return json.dumps(raw_id)
    return None

def forward_frame(url: str, timeout: int, line: str) -> None:
    body = line.encode("utf-8")
    req = urllib.request.Request(
        url, data=body,
        headers={
            "Content-Type": "application/json",
            "Accept": "application/json, text/event-stream",
        },
        method="POST",
    )
    sid = get_session_id()
    if sid:
        req.add_header("Mcp-Session-Id", sid)

    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            status = resp.status
            # Capture session id from response headers
            new_sid = resp.headers.get("Mcp-Session-Id")
            if new_sid:
                set_session_id(new_sid)
            resp_body = resp.read()
        if status == 202 or not resp_body:
            return
        content_type = resp.headers.get("Content-Type", "")
        if "text/event-stream" in content_type:
            resp_body = extract_sse_data(resp_body)
        if not resp_body:
            return
        # May contain multiple frames (one per SSE data: line)
        for frame in resp_body.split(b"\n"):
            frame = frame.strip()
            if frame:
                write_frame(frame)
    except Exception as exc:
        sys.stderr.write(f"mcp-stdio-bridge: request failed: {exc}\n")
        rid = request_id(line)
        if rid is not None:
            msg = str(exc).replace("\\", "\\\\").replace('"', "'")
            err_frame = (
                f'{{"jsonrpc":"2.0","id":{rid},'
                f'"error":{{"code":-32603,"message":"bridge: {msg}"}}}}'
            )
            write_frame(err_frame.encode("utf-8"))

# -- Startup health check ---------------------------------------------------

def wait_for_server(url: str, max_wait: int = 30) -> bool:
    """Poll via a POST with a minimal JSON-RPC ping. HEAD/GET hang on SSE endpoints."""
    deadline = time.monotonic() + max_wait
    attempt = 0
    ping = b'{"jsonrpc":"2.0","id":"ping","method":"initialize","params":{"protocolVersion":"2025-11-25","capabilities":{},"clientInfo":{"name":"healthcheck","version":"1"}}}'
    while time.monotonic() < deadline:
        try:
            req = urllib.request.Request(
                url, data=ping, method="POST",
                headers={"Content-Type": "application/json",
                         "Accept": "application/json, text/event-stream"})
            with urllib.request.urlopen(req, timeout=2) as resp:
                # Capture session id from health check too
                new_sid = resp.headers.get("Mcp-Session-Id")
                if new_sid:
                    set_session_id(new_sid)
                return True
        except Exception:
            pass
        attempt += 1
        if attempt == 5:
            sys.stderr.write(
                f"mcp-stdio-bridge: server not up yet at {url}, waiting...\n"
            )
        time.sleep(0.5)
    return False

# -- Main loop ---------------------------------------------------------------

def main():
    url, timeout, max_concurrent = parse_args()
    if not wait_for_server(url):
        sys.stderr.write(f"mcp-stdio-bridge: no server answering at {url} after 30s\n")
        sys.exit(1)
    sys.stderr.write(f"mcp-stdio-bridge: bridging stdio -> {url}\n")

    pool = ThreadPoolExecutor(max_workers=max_concurrent)
    try:
        for raw_line in sys.stdin.buffer:
            line = raw_line.strip()
            if not line:
                continue
            if len(line) > MAX_LINE_BYTES:
                sys.stderr.write(f"mcp-stdio-bridge: frame exceeds {MAX_LINE_BYTES} bytes, dropped\n")
                continue
            decoded = line.decode("utf-8", errors="replace")
            pool.submit(forward_frame, url, timeout, decoded)
    except KeyboardInterrupt:
        pass
    pool.shutdown(wait=True)
    sys.stderr.write("mcp-stdio-bridge: stdin closed, exiting\n")

if __name__ == "__main__":
    main()
